homeassistant/components/voice_assistant/websocket_api.py

```diff
@@ -46,38 +57,84 @@ async def websocket_run(
-    run_task = hass.async_create_task(
-        TextPipelineRequest(
-            intent_input=msg["intent_input"],
-            conversation_id=msg.get("conversation_id"),
-        ).execute(
-            PipelineRun(
-                hass,
-                connection.context(msg),
-                pipeline,
-                event_callback=lambda event: connection.send_event(
-                    msg["id"], event.as_dict()
+    ws_result: Any | None = None
+
+    intent_input = msg.get("intent_input")
+    if intent_input is None:
+        _LOGGER.debug("Running audio pipeline")
+
+        # Audio pipeline
+        audio_queue: "asyncio.Queue[bytes]" = asyncio.Queue()
+
+        async def stt_stream():
+            while chunk := await audio_queue.get():
+                _LOGGER.debug("Received %s byte(s) of audio", len(chunk))
+                yield chunk
+
+        def handle_binary(_hass, _connection, data: bytes):
+            audio_queue.put_nowait(data)
+
+        handler_id, _unregister_handler = connection.async_register_binary_handler(
+            handle_binary
+        )
+        ws_result = {"handler_id": handler_id}
+
+        run_task = hass.async_create_task(
+            AudioPipelineRequest(
+                stt_metadata=stt.SpeechMetadata(
+                    language=language,
+                    format=stt.AudioFormats.WAV,
+                    codec=stt.AudioCodecs.PCM,
+                    bit_rate=stt.AudioBitRates.BITRATE_16,
+                    sample_rate=stt.AudioSampleRates.SAMPLERATE_16000,
+                    channel=stt.AudioChannels.CHANNEL_MONO,
+                ),
+                stt_stream=stt_stream(),
+                conversation_id=msg.get("conversation_id"),
+            ).execute(
+                PipelineRun(
+                    hass,
+                    connection.context(msg),
+                    pipeline,
+                    event_callback=lambda event: connection.send_event(
+                        msg["id"], event.as_dict()
+                    ),
                 ),
-            ),
-            timeout=timeout,
+                timeout=timeout,
+            )
+        )
+    else:
+        # Text pipeline
+        run_task = hass.async_create_task(
+            TextPipelineRequest(
+                intent_input=msg["intent_input"],
+                conversation_id=msg.get("conversation_id"),
+            ).execute(
+                PipelineRun(
+                    hass,
+                    connection.context(msg),
+                    pipeline,
+                    event_callback=lambda event: connection.send_event(
+                        msg["id"], event.as_dict()
+                    ),
+                ),
+                timeout=timeout,
+            )
         )
-    )
 
     # Cancel pipeline if user unsubscribes
     connection.subscriptions[msg["id"]] = run_task.cancel
 
-    connection.send_result(msg["id"])
+    connection.send_result(msg["id"], ws_result)
```

Comment:
There is no result for a subscription and it's not accessible in the frontend. Only events are supported. Handler ID needs to be included in the run-start event.